### PR TITLE
feat: Add support of TakeOrderedAndProjectExec in Comet

### LINKED ISSUE #89: Add support of TakeOrderedAndProjectExec in Comet
### What is the problem the feature request solves?

`TakeOrderedAndProjectExec` is a common operator in Spark. In Comet, we should support it to increase query coverage.

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #89.

## Rationale for this change

`TakeOrderedAndProjectExec` is a common operator in Spark. In Comet, we should support it to increase query coverage.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+    if (scanTypes.length == outputAttributes.length) {
+      scanBuilder.addAllFields(scanTypes.asJava)
+
+      val sortOrders = sortOrder.map(exprToProto(_, child.output))
+
+      if (sortOrders.forall(_.isDefined)) {
+        val sortBuilder = OperatorOuterClass.Sort.newBuilder()
+        sortBuilder.addAllSortOrders(sortOrders.map(_.get).asJava)
+        sortBuilder.setFetch(limit)
```
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala:211]: Yes.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+      "number of partitions")) ++ readMetrics ++ writeMetrics
+
+  private lazy val serializer: Serializer =
+    new UnsafeRowSerializer(child.output.size, longMetric("dataSize"))
+
+  // Exposed for testing.
+  lazy val orderingSatisfies: Boolean =
+    SortOrder.orderingSatisfies(child.outputOrdering, sortOrder)
+
+  protected override def doExecuteColumnar(): RDD[ColumnarBatch] = {
```
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala:74]: Yes, `CometExec` provides default `executeCollect` implementation.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+          .newBuilder()
+          .addChildren(topK)
+        Some(opBuilder.setProjection(projectBuilder).build())
+      } else {
+        None
+      }
+    }
+  }
+
+  def getLimitNativePlan(outputAttributes: Seq[Attribute], limit: Int): Option[Operator] = {
```
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometTakeOrderedAndProjectExec.scala:165]: Okay. I will move them to a separate place.

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala [hunk 4]
```diff
@@ -320,6 +320,20 @@ class CometSparkSessionExtensions
               c
           }
 
+        case s: TakeOrderedAndProjectExec
+            if isCometNative(s.child) && isCometOperatorEnabled(conf, "takeOrderedAndProjectExec")
+              && isCometShuffleEnabled(conf) &&
+              CometTakeOrderedAndProjectExec.isSupported(s.projectList, s.sortOrder, s.child) =>
+          // TODO: support offset for Spark 3.4
```
[the remark this answers is not in the document]
Comment (sunchao) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:327]: We can probably fallback to Spark when `offset > 0`?

[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:327]: Yea, we need to do fallback for `offset > 0`.

[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:328]: Let me do fallback in this PR. I will address offset > 0 case in follow ups.

**Comment by viirya** [comment]
I was wondering why there is test failure as I didn't see it locally. But after rebased with `main` branch, I can reproduce it locally. Looking into which commit causes it...

**Comment by viirya** [comment]
Found the cause. We need `CometTakeOrderedAndProjectExec` listed in a pattern match list. Because we removed it from previous contributed patches, forgot to add it back.

**Review by sunchao — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Merged. Thanks.

**Pull request merged**